client/components/site-preview-link/index.tsx

```diff
@@ -0,0 +1,124 @@
+import { useCreateSitePreviewLink } from './use-create-site-preview-link';
+import { useDeleteSitePreviewLink } from './use-delete-site-preview-link';
+import { useSitePreviewLinks } from './use-site-preview-links';
+import type { SiteId } from 'calypso/types';
+
+interface SitePreviewLinkProps {
+	siteId: SiteId;
+	siteUrl: string;
+	source: 'launch-settings' | 'privacy-settings' | 'smp-modal';
+	disabled?: boolean;
+}
+
+const HelpText = styled.p( {
+	display: 'block',
+	margin: '5px 0',
+	fontSize: '0.875rem',
+	fontStyle: 'italic',
+	fontWeight: 400,
+	color: 'var(--color-text-subtle)',
+} );
+
+const NOTICE_OPTIONS = {
+	duration: 5000,
+};
+
+function useDispatchNotice() {
+	const dispatch = useDispatch();
+
+	return {
+		showSuccessNotice: ( message: string ) => dispatch( successNotice( message, NOTICE_OPTIONS ) ),
+		showErrorNotice: ( message: string ) => dispatch( errorNotice( message, NOTICE_OPTIONS ) ),
+	};
+}
+
+export default function SitePreviewLink( {
+	siteId,
+	siteUrl,
+	disabled = false,
+	source,
+}: SitePreviewLinkProps ) {
+	const translate = useTranslate();
+	const { showSuccessNotice, showErrorNotice } = useDispatchNotice();
+	const [ checked, setChecked ] = useState( false );
+
+	const { data: previewLinks, isLoading: isFirstLoading } = useSitePreviewLinks( {
+		siteId,
+		onSuccess: ( linksResponse ) => {
+			const validLinksLength = linksResponse?.filter( ( link ) => ! link.isRemoving ).length || 0;
+			setChecked( validLinksLength > 0 );
+		},
+	} );
+
+	const { createLink, isLoading: isCreating } = useCreateSitePreviewLink( {
+		siteId,
+		onSuccess: () => {
+			showSuccessNotice( translate( 'Preview link created successfully!' ) );
+			recordTracksEvent( 'calypso_site_preview_link_created', { source } );
+		},
+		onError: () => {
+			showErrorNotice( translate( 'Error creating the preview link.' ) );
+			recordTracksEvent( 'calypso_site_preview_link_created_error', { source } );
+		},
+	} );
+
+	const { deleteLink, isLoading: isDeleting } = useDeleteSitePreviewLink( {
+		siteId,
+		onSuccess: () => {
+			showSuccessNotice( translate( 'Preview link removed successfully!' ) );
+			recordTracksEvent( 'calypso_site_preview_link_deleted', { source } );
+		},
+		onError: () => {
+			showErrorNotice( translate( 'Error removing the preview link.' ) );
+			recordTracksEvent( 'calypso_site_preview_link_deleted_error', { source } );
+		},
+	} );
+
+	function onChange( checkedValue: boolean ) {
+		setChecked( checkedValue );
+		if ( checkedValue ) {
+			createLink();
+		} else {
+			previewLinks?.map( ( { code } ) => {
+				deleteLink( code );
+			} );
+		}
+	}
+
+	const checkedAndEnabled = checked && ! disabled;
+	const isBusy = isFirstLoading || isCreating || isDeleting;
+
+	if ( isFirstLoading ) {
+		return <Spinner />;
+	}
+	return (
+		<div>
+			<ToggleControl
+				label={ translate( 'Create a preview link.' ) }
+				checked={ checkedAndEnabled }
+				onChange={ onChange }
```

Comment:
I noticed one more thing here - when the site is in "Coming Soon" mode, I can click on "Discourage search engines from indexing this site" and it will automatically switch the site to "Public". It would be reasonable to support it and vice-versa too. However, we can't do it as a preview link can't be generated if the site is not Coming Soon, and the site won't be Coming Soon until the user clicks "Save settings".

Maybe we should go with the initial idea and glue the preview links block below the "Privacy" block, and display it only after site becomes Coming Soon and when changes are saved?

Reply:
> When I launch the site, the toggle is disabled, but I can still click on it and trigger a loading indicator. Should we prevent that?

I think it's a really weird case. I am usually redirected to buy a domain screen.

> Also, it reloads when I unfocus and focus on the site e.g. by switching to another tab and switching to the Calypso tab again. After having the tab open for a while I see a bunch of requests to the preview-link endpoint. This occurs also when the control is in an active state.

I'm not seeing the spinner for this case. The requests are made to check the latest state of the page.
If it's an error it retries the requests.
We could update the backend to return an empty array or we can disable the retries on error for this react query.

> I noticed one more thing here - when the site is in "Coming Soon" mode, I can click on "Discourage search engines from indexing this site" and it will automatically switch the site to "Public". It would be reasonable to support it and vice-versa too. However, we can't do it as a preview link can't be generated if the site is not Coming Soon, and the site won't be Coming Soon until the user clicks "Save settings".
> Maybe we should go with the initial idea and glue the preview links block below the "Privacy" block, and display it only after site becomes Coming Soon and when changes are saved?

Yes, I like showing the toggle only after the settings are updated. Thanks for reporting this !! 🙌